### PR TITLE
raspberrypi: implement setsockopt(SOL_SOCKET, SO_REUSEADDR)

### LINKED ISSUE #9081: httpserver simpletest fails (cp9 on Pico W)
### CircuitPython version

```python
dafruit CircuitPython 9.0.0 on 2024-03-19; Raspberry Pi Pico W with rp2040
```

### Code/REPL

```python
import socketpool
import wifi

from adafruit_httpserver import Server, Request, Response

ssid = "Unintended Consequences"
password = "xxxx"

print("Connecting to", ssid)
wifi.radio.connect(ssid, password)
print("Connected to", ssid)

pool = socketpool.SocketPool(wifi.radio)
server = Server(pool, "/static", debug=True)


@server.route("/")
def base(request: Request):
    """
    Serve a default static plain text message.
    """
    return Response(request, "Hello from the CircuitPython HTTP Server!")


server.serve_forever(str(wifi.radio.ipv4_address))
```

### Behavior

code.py output:
Connecting to Unintended Consequences
Connected to Unintended Consequences
Traceback (most recent call last):
  File "code.py", line 25, in <module>
  File "adafruit_httpserver/server.py", line 185, in serve_forever
  File "adafruit_httpserver/server.py", line 228, in start
  File "adafruit_httpserver/server.py", line 207, in _create_server_socket
OSError: [Errno 95] EOPNOTSUPP

### Description

Test code taken from first example here: https://docs.circuitpython.org/projects/httpserver/en/latest/examples.html

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by jepler**
untested

edit by @dhalbert:
Fixes #9081.

### CODE REVIEW

File: ports/raspberrypi/common-hal/socketpool/Socket.c [hunk 1]
```diff
@@ -1184,15 +1184,35 @@ void common_hal_socketpool_socket_settimeout(socketpool_socket_obj_t *self, uint
...
+            }
+            break;
+
+        case SOCKETPOOL_SOL_SOCKET:
+            switch (optname) {
+                case SOCKETPOOL_SO_REUSEADDR:
+                    if (enable) {
+                        ip_set_option(self->pcb.ip, SOF_REUSEADDR);
+                    } else {
+                        ip_set_option(self->pcb.ip, SOF_REUSEADDR);
```
Comment (anecdata) [inline ports/raspberrypi/common-hal/socketpool/Socket.c:1210]: Is this intended to be `set` for both conditions?

**Comment by anecdata** [comment]
`Adafruit CircuitPython 9.0.0-2-g1ad5a73eb5 on 2024-03-25; Raspberry Pi Pico W with rp2040`
+
latest release `adafruit_httpserver` 4.5.6

...works fine, no exceptions.

A number of resets of the server didn't yield any exceptions. We can test the no-reuse case pending the question above. I'm not sure how to reliably trigger the address in use exception though.

**Comment by anecdata** [comment]
With today's artifact and same latest release of the adafruit_httpserver library, I get `OSError: [Errno 112] EADDRINUSE` upon `bind` by default when the server is interrupted (^C) and restarted (^D). This code is in server.py:
```py
        if implementation.version >= (9,) or implementation.name != "circuitpython":
            sock.setsockopt(socket_source.SOL_SOCKET, socket_source.SO_REUSEADDR, 1)
```

But if I change it to `sock.setsockopt(socket_source.SOL_SOCKET, socket_source.SO_REUSEADDR, 0)`, then there is no `EADDRINUSE` exception. Backwards (I think?), but not sure why.

**Review by anecdata — APPROVED** [review]
`.SO_REUSEADDR, 1` now prevents `EADDRINUSE` and allows the server to immediately re-bind, and `.SO_REUSEADDR, 0` raises `EADDRINUSE` within the timeout. Thanks, @jepler !

**Comment by jepler** [comment]
really appreciated the testing @anecdata !

**Pull request merged**